File: app/src/main/java/rogue_opcode/TextSE.java
// TextSE.java
// Text-centric ScreenElement subclass.
//
// Copyright © 2012-2017 Christopher R. Tooley
// This software is part of the Rogue-Opcode game framework. It is distributable
// under the terms of a modified MIT License. You should have received a copy of
// the license in the file LICENSE.md. If not, see:
// <http://code.google.com/p/rogue-op/wiki/LICENSE>
//
// Formatting:
// 80 cols ; tabwidth 4
// //////////////////////////////////////////////////////////////////////////////


package rogue_opcode;

import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.LinearGradient;
import android.graphics.Paint;
import android.graphics.RectF;
import android.graphics.Region;
import android.graphics.Shader;
import android.graphics.Typeface;
import android.text.DynamicLayout;
import android.text.Layout;
import android.text.TextPaint;

import rogue_opcode.GameProc.TouchState;
import rogue_opcode.geometrics.XYf;


public class TextSE extends ScreenElement
{
	private static final long serialVersionUID = 1L;
	static TextPaint sSmallTextPaint;
	static TextPaint sLabelTextPaint;
	static TextPaint sFillPaint;
	static int sTextMargin = 10;

	DynamicLayout mTextDL;
	String mTitle;
	int mWidth; //overall width - includes icon width
	int mIconWidth;
	int mHeight;
	int mTextBottomMargin;
	boolean mFixedHeight;
	RectF mDisplayRect;
	RectF mTextClipRect;

	boolean mInEditMode;
	boolean mEditable;
	boolean mDirtyFlag; //set only if USER edited the text

	//Variable-Height constructor
	public TextSE(String pText, float pX, float pY, int pWidth)
	{
		super(0);
		initialize(pText, pX, pY, pWidth, -1);
	}

	//Fixed-Height constructor
	public TextSE(String pText, float pX, float pY, int pWidth, int pHeight)
	{
		super(0);
		initialize(pText, pX, pY, pWidth, pHeight);
	}

	public boolean Dirty()
	{
		boolean tDirty = mDirtyFlag;
		Dirty(false);

		return tDirty;
	}

	void Dirty(boolean pDirty)
	{
		mDirtyFlag = pDirty;
	}

	void initialize(String pText, float pX, float pY, int pWidth, int pHeight)
	{
		mHeight = pHeight;

		mFixedHeight = (mHeight != -1);

		mWidth = pWidth;
		mIconWidth = 0;

		mTextBottomMargin = 0;

		mPos.x = pX;
		mPos.y = pY;


		if(sSmallTextPaint == null)
		{
			sSmallTextPaint = new TextPaint();
			sSmallTextPaint.setTypeface(Typeface.createFromAsset(
				GameProc.sOnly.getAssets(), "fonts/aispec.ttf")); //TODO - should be a call from outside of rogue-op
			sSmallTextPaint.setColor(Color.BLACK);
			sSmallTextPaint.setAlpha(0xff);
			sSmallTextPaint.setAntiAlias(true);
			sSmallTextPaint.setTextSize(16);
			sSmallTextPaint.setTextAlign(Paint.Align.LEFT);
		}

		if(sLabelTextPaint == null)
		{
			sLabelTextPaint = new TextPaint();
			sLabelTextPaint.setColor(0xff8ae234);
			sLabelTextPaint.setAntiAlias(true);
			sLabelTextPaint.setTextSize(16);
			sLabelTextPaint.setTextSkewX(-.25f);
			sLabelTextPaint.setTextAlign(Paint.Align.LEFT);
		}

		if(sFillPaint == null)
		{
			sFillPaint = new TextPaint();
			sFillPaint.setColor(Color.GRAY);
			sFillPaint.setAlpha(0xbb);
			sFillPaint.setAntiAlias(true);
			sFillPaint.setShader(new LinearGradient(0, 0, 0, 20, Color.WHITE,
				Color.GRAY, Shader.TileMode.CLAMP));
		}

		mTextPaint = sSmallTextPaint;

		set_text(pText);

		mInEditMode = false;
		mEditable = false;
		mDirtyFlag = false;

		ZDepth(100);
	}

	//Virtual width (not Physical width)
	@Override
	public int Width()
	{
		return (mWidth);
	}

	@Override
	public int Height()
	{
		return (mHeight);
	}

	public void AddIcon(int pGRID, int pIconWidth)
	{
		mIconWidth = (pIconWidth);
		SetCurrentGR(pGRID);
		set_text(mText);
	}

	public void SetTextBottomMargin(int pMargin)
	{
		mTextBottomMargin = pMargin;
		set_text(mText);
	}

	@Override
	public void Text(String pText)
	{
		set_text(pText);
	}

	public String Title()
	{
		return (mTitle == null) ? "" : mTitle;
	}

	public void Title(String pTitle)
	{
		mTitle = pTitle;
	}

	public void SetTextPaint(TextPaint pTextPaint)
	{
		mTextPaint = pTextPaint;

		set_text(Text());
	}

	protected void set_text(String pText)
	{
		mText = pText;

		try
		{
			mTextDL =
				new DynamicLayout(mText, (TextPaint)mTextPaint,
					mWidth - (sTextMargin * 2) - (sTextMargin + mIconWidth),
					Layout.Alignment.ALIGN_NORMAL, 1.0f, 0.0f, false);
		}
		catch(Exception e)
		{
			//The text and icon don't fit into this mWidth
		}
		if(!mFixedHeight)
		{
			mHeight =
				mTextDL.getLineTop(mTextDL.getLineCount()) + (sTextMargin * 2);

			if(mHeight < ((sTextMargin * 2) + mIconWidth))
				mHeight = (sTextMargin * 2) + mIconWidth;
		}

		mDisplayRect = new RectF(0, 0, mWidth, mHeight);
		mTextClipRect = new RectF(0, 0, mWidth, mHeight - mTextBottomMargin);
	}

	public boolean WithinRange(XYf pPoint)
	{
		return ((pPoint.x > mPos.x) && (pPoint.x < mPos.x + mWidth) &&
				(pPoint.y > mPos.y) && (pPoint.y < mPos.y + mHeight));
	}

	public void Editable(boolean pEditable)
	{
		mEditable = pEditable;
	}

	@Override
	public void Visible(boolean pVisible)
	{
		super.Visible(pVisible);

		if(mInEditMode)
		{
			GameProc.sOnly.HideTextEditor(this);
			mInEditMode = false;
		}
	}

	@Override
	public void Update()
	{
		if(mEditable)
		{
			try
			{
				if(mInEditMode)
				{
					if(GameProc.sOnly.mEditTextParams.mCurrentTE == this)
						set_text(GameProc.sOnly.mEditTextParams.mEdit.getText()
																		.toString());

					if(GameProc.sOnly.mTouchState.Is(TouchState.SINGLE_TAP))
					{
						if(!WithinRange(new XYf(
							GameProc.sOnly.mTouchState.GetMainX(),
							GameProc.sOnly.mTouchState.GetMainY())))
						{
							mInEditMode = false;
							Dirty(true);
							GameProc.sOnly.HideTextEditor(this);
						}
					}
				}
				else
				{
					if(GameProc.sOnly.mTouchState.Is(TouchState.SINGLE_TAP))
					{
						if(Visible())
						{
							if(WithinRange(new XYf(
								GameProc.sOnly.mTouchState.GetMainX(),
								GameProc.sOnly.mTouchState.GetMainY())))
							{
								mInEditMode = true;
								GameProc.sOnly.ShowTextEditor(this, mPos,
									mWidth, mHeight);
							}
						}
					}
				}
			}
			catch(Exception e)
			{
			}
		}
	}


	@Override
	public void Draw()
	{
		Canvas tCanvas = AnimatedView.sCurrentCanvas;
		float tX = mPos.x;// * AnimatedView.sOnly.mPreScaler;
		float tY = mPos.y;// * AnimatedView.sOnly.mPreScaler;

		tCanvas.save();
		tCanvas.scale(AnimatedView.sOnly.mPreScaler,
			AnimatedView.sOnly.mPreScaler);

		tCanvas.translate(tX, tY);
		tCanvas.drawRoundRect(mDisplayRect, 10, 10, sFillPaint);
		tCanvas.drawText(Title(), 10, -2, sLabelTextPaint);
		tCanvas.translate(sTextMargin + sTextMargin + mIconWidth, sTextMargin);
		tCanvas.clipRect(mTextClipRect, Region.Op.REPLACE);
		mTextDL.draw(tCanvas);
		tCanvas.restore();

		tCanvas.save();
		int tIconOffset =
			(int)(AnimatedView.sOnly.mPreScaler * ((mIconWidth / 2) + sTextMargin));
		tCanvas.translate(tIconOffset, tIconOffset);
		super.Draw();
		tCanvas.restore();
	}
}

File: lib/src/main/java/rogue_opcode/TextSE.kt
// TextSE.kt
// Text-centric ScreenElement subclass.
//
// Copyright © 2012-2017 Christopher R. Tooley
// This software is part of the Rogue-Opcode game framework. It is distributable
// under the terms of a modified MIT License. You should have received a copy of
// the license in the file LICENSE.md. If not, see:
// <https://github.com/IonoclastBrigham/RogueOp/blob/master/LICENSE.md>
//
// Formatting:
// 80 cols ; tabwidth 4
// //////////////////////////////////////////////////////////////////////////////


package rogue_opcode

import android.graphics.*
import android.text.DynamicLayout
import android.text.Layout
import android.text.TextPaint
import rogue_opcode.geometrics.XYf


class TextSE(pText: String, pX: Float, pY: Float, pWidth: Int, pHeight: Int = -1)
	: ScreenElement(0) {

	companion object {
		private val serialVersionUID = 1L

		private var sLabelTextPaint = TextPaint().apply {
			color = Color.argb(0xff, 0x8a, 0xe2, 0x34)
			isAntiAlias = true
			textSize = 16f
			textSkewX = -.25f
			textAlign = Paint.Align.LEFT
		}
		private var sSmallTextPaint = TextPaint().apply {
			//TODO - should be a call from outside of rogue-op
			typeface = Typeface.createFromAsset(GameProc.sOnly.assets,
			                                    "fonts/aispec.ttf")
			color = Color.BLACK
			alpha = 0xff
			isAntiAlias = true
			textSize = 16f
			textAlign = Paint.Align.LEFT
		}
		private var sFillPaint = TextPaint().apply {
			color = Color.GRAY
			alpha = 0xbb
			isAntiAlias = true
			shader = LinearGradient(0f, 0f, 0f, 20f,
			                        Color.WHITE, Color.GRAY,
			                        Shader.TileMode.CLAMP)
		}
		private var sTextMargin = 10
	}

	private lateinit var mTextDL: DynamicLayout
	private var mTitle: String? = null
	private var mWidth: Int = 0 //overall width - includes icon width
	private var mIconWidth: Int = 0
	private var mHeight: Int = 0
	private var mTextBottomMargin: Int = 0
	private var mFixedHeight: Boolean = false
	private lateinit var mDisplayRect: RectF
	private lateinit var mTextClipRect: RectF

	private var mInEditMode: Boolean = false
	private var mEditable: Boolean = false
	private var mDirtyFlag: Boolean = false //set only if USER edited the text

	init {
		mHeight = pHeight
		mFixedHeight = (mHeight != -1)

		mWidth = pWidth
		mIconWidth = 0
		mTextBottomMargin = 0

		mPos.x = pX
		mPos.y = pY

		mTextPaint = sSmallTextPaint

		set_text(pText)

		mInEditMode = false
		mEditable = false
		mDirtyFlag = false

		ZDepth(100f)
	}

	fun Dirty(): Boolean {
		val tDirty = mDirtyFlag
		Dirty(false)

		return tDirty
	}

	internal fun Dirty(pDirty: Boolean) {
		mDirtyFlag = pDirty
	}

	//Virtual width (not Physical width)
	override fun Width(): Int {
		return mWidth
	}

	override fun Height(): Int {
		return mHeight
	}

	fun AddIcon(pGRID: Int, pIconWidth: Int) {
		mIconWidth = pIconWidth
		SetCurrentGR(pGRID)
		set_text(mText ?: "")
	}

	fun SetTextBottomMargin(pMargin: Int) {
		mTextBottomMargin = pMargin
		set_text(mText ?: "")
	}

	override fun Text(pText: String) {
		set_text(pText)
	}

	fun Title() = mTitle ?: ""

	fun Title(pTitle: String) {
		mTitle = pTitle
	}

	fun SetTextPaint(pTextPaint: TextPaint) {
		mTextPaint = pTextPaint

		set_text(mText ?: "")
	}

	private fun set_text(pText: String) {
		mText = pText

		try {
			mTextDL = DynamicLayout(mText, mTextPaint as TextPaint,
					mWidth - sTextMargin * 2 - (sTextMargin + mIconWidth),
					Layout.Alignment.ALIGN_NORMAL, 1.0f, 0.0f, false)
		} catch(e: Exception) {
			//The text and icon don't fit into this mWidth
		}

		if(!mFixedHeight) {
			mHeight = mTextDL.getLineTop(mTextDL.lineCount) + sTextMargin * 2

			if(mHeight < sTextMargin * 2 + mIconWidth)
				mHeight = sTextMargin * 2 + mIconWidth
		}

		mDisplayRect = RectF(0f, 0f, mWidth.toFloat(), mHeight.toFloat())
		mTextClipRect = RectF(0f, 0f, mWidth.toFloat(), (mHeight - mTextBottomMargin).toFloat())
	}

	fun WithinRange(pPoint: XYf): Boolean {
		return pPoint.x > mPos.x && pPoint.x < mPos.x + mWidth &&
				pPoint.y > mPos.y && pPoint.y < mPos.y + mHeight
	}

	fun Editable(pEditable: Boolean) {
		mEditable = pEditable
	}

	override var Visible = false
		set(pVisible) {
			super.Visible = pVisible

			if(mInEditMode) {
				GameProc.sOnly.HideTextEditor(this)
				mInEditMode = false
			}
		}

	override fun Update() {
		if(!mEditable) return

		val tProc = GameProc.sOnly
		if(mInEditMode) {
			if(tProc.mEditTextParams?.mCurrentTE === this) {
				set_text(tProc.mEditTextParams?.mEdit?.text.toString())
			}

			if(TouchState.Is(TouchState.SINGLE_TAP)) {
				if(!WithinRange(TouchState.MainTouchPos())) {
					mInEditMode = false
					Dirty(true)
					tProc.HideTextEditor(this)
				}
			}
		} else if(TouchState.Is(TouchState.SINGLE_TAP) && Visible) {
			if (WithinRange(TouchState.MainTouchPos())) {
				mInEditMode = true
				tProc.ShowTextEditor(this, mPos, mWidth, mHeight)
			}
		}
	}


	override fun Draw(pCanvas: Canvas, pPreScaler: Float, pDefaultPaint: Paint) {
		val tX = mPos.x// * pPreScaler;
		val tY = mPos.y// * pPreScaler;

		pCanvas.save()
		pCanvas.scale(pPreScaler, pPreScaler)

		pCanvas.translate(tX, tY)
		pCanvas.drawRoundRect(mDisplayRect, 10f, 10f, sFillPaint)
		pCanvas.drawText(Title(), 10f, -2f, sLabelTextPaint)
		pCanvas.translate((sTextMargin + sTextMargin + mIconWidth).toFloat(),
		                  sTextMargin.toFloat())
		pCanvas.clipRect(mTextClipRect, Region.Op.REPLACE)
		mTextDL.draw(pCanvas)
		pCanvas.restore()

		pCanvas.save()
		val tIconOffset = (pPreScaler * (mIconWidth / 2 + sTextMargin)).toInt()
		pCanvas.translate(tIconOffset.toFloat(), tIconOffset.toFloat())
		super.Draw(pCanvas, pPreScaler, pDefaultPaint)
		pCanvas.restore()
	}
}
